tests/unit/s2n_server_hello_retry_test.c

```diff
@@ -412,6 +412,100 @@ int main(int argc, char **argv)
         EXPECT_SUCCESS(s2n_cert_chain_and_key_free(tls13_chain_and_key));
     }
 
+    /* Send and receive Hello Retry Request messages, test for non blocking client hello callback */
+    {
+        struct s2n_config *server_config;
+        struct s2n_config *client_config;
+
+        struct s2n_connection *server_conn;
+        struct s2n_connection *client_conn;
+
+        struct s2n_cert_chain_and_key *tls13_chain_and_key;
+        char tls13_cert_chain[S2N_MAX_TEST_PEM_SIZE] = {0};
+        char tls13_private_key[S2N_MAX_TEST_PEM_SIZE] = {0};
+
+        EXPECT_NOT_NULL(server_config = s2n_config_new());
+        EXPECT_NOT_NULL(client_config = s2n_config_new());
+
+        EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+        EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+
+        EXPECT_SUCCESS(s2n_config_set_cipher_preferences(client_config, "default_tls13"));
+        EXPECT_SUCCESS(s2n_config_set_cipher_preferences(server_config, "default_tls13"));
+
+        EXPECT_NOT_NULL(tls13_chain_and_key = s2n_cert_chain_and_key_new());
+        EXPECT_SUCCESS(s2n_read_test_pem(S2N_ECDSA_P384_PKCS1_CERT_CHAIN, tls13_cert_chain, S2N_MAX_TEST_PEM_SIZE));
+        EXPECT_SUCCESS(s2n_read_test_pem(S2N_ECDSA_P384_PKCS1_KEY, tls13_private_key, S2N_MAX_TEST_PEM_SIZE));
+        EXPECT_SUCCESS(s2n_cert_chain_and_key_load_pem(tls13_chain_and_key, tls13_cert_chain, tls13_private_key));
+        EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(client_config, tls13_chain_and_key));
+        EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(server_config, tls13_chain_and_key));
+
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, server_config));
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, client_config));
+        struct client_hello_context client_hello_ctx = {.invocations = 0 };
+        EXPECT_SUCCESS(s2n_config_set_client_hello_cb(server_config, client_hello_detect_duplicate_calls, &client_hello_ctx));
+        EXPECT_SUCCESS(s2n_config_set_client_hello_cb_mode(server_config,
+            S2N_CLIENT_HELLO_CB_NONBLOCKING));
+
+        EXPECT_SUCCESS(s2n_connection_set_all_protocol_versions(server_conn, S2N_TLS13));
+        EXPECT_SUCCESS(s2n_connection_set_all_protocol_versions(client_conn, S2N_TLS13));
+
+        /* Force HRR path by sending an empty list of keyshares */
+        EXPECT_SUCCESS(s2n_connection_set_keyshare_by_name_for_testing(client_conn, "none"));
+
+        /* Send the first CH message */
+        EXPECT_SUCCESS(s2n_client_hello_send(client_conn));
+        EXPECT_SUCCESS(s2n_stuffer_copy(&client_conn->handshake.io, &server_conn->handshake.io,
+                                        s2n_stuffer_data_available(&client_conn->handshake.io)));
```

Comment:
Why send the messages manually instead of calling s2n_negotiate? If you don't want to do the full handshake, there's also https://github.com/aws/s2n-tls/blob/main/tests/testlib/s2n_testlib.h#L156-L157

Reply:
these sets follow the same mechanics as rest of this file. to keep this PR short, if it's fine with you would rather follow the same approach.

Reply:
Unlike the other hello retry tests, you're not really interested in the details of the retry. You just want to make sure that the second ClientHello doesn't trigger the callback. A more reliable way of testing that is to NOT introduce the possibility that you send the messages wrong, and just let S2N do it for you. Focus on what you actually want to verify.

Reply:
sure, done.